Add DEBUG infos to profile cuDNN convolutions.

notoraptor opened the pull request:
This PR adds profiling in C DEBUG mode for cuDNN convolutions.

Currently, profiling is added only for `dnn_fwd`, but I am not sure if current code efficiently measures execution time. Even with `CUDA_LAUNCH_BLOCKING=1`, some executions take `0 milliseconds` (see below).

I have added a special test class (`TestDnnConv2DRuntimeAlgorithmsWithBigInputs`) to test with "big" shapes so that we can expect more computation time. Here is the result (tested on Windows 10 64 bits with cuDNN V7, flag `cmodule.debug=True` and `CUDA_LAUNCH_BLOCKING=1`):
```
(python2) C:\Users\notoraptor\mila\dev\git\theano>nosetests -xvs theano\gpuarray\tests\test_dnn.py:TestDnnConv2DRuntimeAlgorithmsWithBigInputs.test_fwd_runtime_algorithms
Using cuDNN version 7001 on context None
Mapped name None to device cuda: GeForce GTX 1050 (0000:01:00.0)
theano.gpuarray.tests.test_dnn.TestDnnConv2DRuntimeAlgorithmsWithBigInputs.test_fwd_runtime_algorithms('time_once',) ... (using CUDNN_CONVOLUTION_FWD_ALGO_FFT[T] (timed), ws:9445376, hash:FWD|GPU#0000:01:00.0 -t -g 1 -dim 12,4,128,128,65536,16384,128,1 -filt 5,4,64,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1 [unaligned])
        (selected fastest fwd algo in 234 milliseconds)
        (ran fwd algo in 16 milliseconds)
(using CUDNN_CONVOLUTION_FWD_ALGO_FFT[T] (timed)(reused), ws:9445376, hash:)
        (ran fwd algo in 0 milliseconds)
        (ran 2 fwd computations in 16 milliseconds (average: 8 milliseconds per call))
(using CUDNN_CONVOLUTION_FWD_ALGO_FFT[T] (timed)(reused), ws:9445376, hash:)
        (ran fwd algo in 0 milliseconds)
        (ran 3 fwd computations in 16 milliseconds (average: 5.33333 milliseconds per call))
(using CUDNN_CONVOLUTION_FWD_ALGO_FFT[T] (timed)(reused), ws:9445376, hash:)
        (ran fwd algo in 0 milliseconds)
        (ran 4 fwd computations in 16 milliseconds (average: 4 milliseconds per call))
(using CUDNN_CONVOLUTION_FWD_ALGO_FFT[T] (timed)(reused), ws:9445376, hash:)
        (ran fwd algo in 0 milliseconds)
        (ran 5 fwd computations in 16 milliseconds (average: 3.2 milliseconds per call))
ok
theano.gpuarray.tests.test_dnn.TestDnnConv2DRuntimeAlgorithmsWithBigInputs.test_fwd_runtime_algorithms('guess_once',) ... (using CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM , ws:65536, hash:FWD|GPU#0000:01:00.0 -g 1 -dim 12,4,128,128,65536,16384,128,1 -filt 5,4,64,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1)
        (selected best suited fwd algo in 0 milliseconds)
        (ran fwd algo in 78 milliseconds)
(using CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM (reused), ws:65536, hash:)
        (ran fwd algo in 78 milliseconds)
        (ran 2 fwd computations in 156 milliseconds (average: 78 milliseconds per call))
(using CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM (reused), ws:65536, hash:)
        (ran fwd algo in 63 milliseconds)
        (ran 3 fwd computations in 219 milliseconds (average: 73 milliseconds per call))
(using CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM (reused), ws:65536, hash:)
        (ran fwd algo in 78 milliseconds)
        (ran 4 fwd computations in 297 milliseconds (average: 74.25 milliseconds per call))
(using CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM (reused), ws:65536, hash:)
        (ran fwd algo in 62 milliseconds)
        (ran 5 fwd computations in 359 milliseconds (average: 71.8 milliseconds per call))
ok
theano.gpuarray.tests.test_dnn.TestDnnConv2DRuntimeAlgorithmsWithBigInputs.test_fwd_runtime_algorithms('time_on_shape_change',) ... (using CUDNN_CONVOLUTION_FWD_ALGO_FFT[T] (timed), ws:9445376, hash:FWD|GPU#0000:01:00.0 -t -g 1 -dim 12,4,128,128,65536,16384,128,1 -filt 5,4,64,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1)
        (selected fastest fwd algo in 203 milliseconds)
        (ran fwd algo in 0 milliseconds)
(using CUDNN_CONVOLUTION_FWD_ALGO_FFT[T] (timed)(cache), ws:9445376, hash:FWD|GPU#0000:01:00.0 -t -g 1 -dim 12,4,128,128,65536,16384,128,1 -filt 5,4,64,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1)
        (ran fwd algo in 0 milliseconds)
        (ran 2 fwd computations in 0 milliseconds (average: 0 milliseconds per call))
(using CUDNN_CONVOLUTION_FWD_ALGO_FFT[T] (timed)(cache), ws:9445376, hash:FWD|GPU#0000:01:00.0 -t -g 1 -dim 12,4,128,128,65536,16384,128,1 -filt 5,4,64,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1)
        (ran fwd algo in 0 milliseconds)
        (ran 3 fwd computations in 0 milliseconds (average: 0 milliseconds per call))
(using CUDNN_CONVOLUTION_FWD_ALGO_FFT[T] (timed)(cache), ws:9445376, hash:FWD|GPU#0000:01:00.0 -t -g 1 -dim 12,4,128,128,65536,16384,128,1 -filt 5,4,64,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1)
        (ran fwd algo in 0 milliseconds)
        (ran 4 fwd computations in 0 milliseconds (average: 0 milliseconds per call))
(using CUDNN_CONVOLUTION_FWD_ALGO_FFT[T] (timed)(cache), ws:9445376, hash:FWD|GPU#0000:01:00.0 -t -g 1 -dim 12,4,128,128,65536,16384,128,1 -filt 5,4,64,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1)
        (ran fwd algo in 0 milliseconds)
        (ran 5 fwd computations in 0 milliseconds (average: 0 milliseconds per call))
(using CUDNN_CONVOLUTION_FWD_ALGO_FFT[T] (timed), ws:37502976, hash:FWD|GPU#0000:01:00.0 -t -g 1 -dim 12,4,256,256,262144,65536,256,1 -filt 5,4,32,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1)
        (selected fastest fwd algo in 1015 milliseconds)
        (selected 2 fwd algos in 1218 milliseconds (average: 609 milliseconds per selection))
        (ran fwd algo in 16 milliseconds)
        (ran 6 fwd computations in 16 milliseconds (average: 2.66667 milliseconds per call))
(using CUDNN_CONVOLUTION_FWD_ALGO_FFT[T] (timed)(cache), ws:37502976, hash:FWD|GPU#0000:01:00.0 -t -g 1 -dim 12,4,256,256,262144,65536,256,1 -filt 5,4,32,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1)
        (ran fwd algo in 0 milliseconds)
        (ran 7 fwd computations in 16 milliseconds (average: 2.28571 milliseconds per call))
(using CUDNN_CONVOLUTION_FWD_ALGO_FFT[T] (timed)(cache), ws:37502976, hash:FWD|GPU#0000:01:00.0 -t -g 1 -dim 12,4,256,256,262144,65536,256,1 -filt 5,4,32,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1)
        (ran fwd algo in 0 milliseconds)
        (ran 8 fwd computations in 16 milliseconds (average: 2 milliseconds per call))
(using CUDNN_CONVOLUTION_FWD_ALGO_FFT[T] (timed)(cache), ws:37502976, hash:FWD|GPU#0000:01:00.0 -t -g 1 -dim 12,4,256,256,262144,65536,256,1 -filt 5,4,32,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1)
        (ran fwd algo in 0 milliseconds)
        (ran 9 fwd computations in 16 milliseconds (average: 1.77778 milliseconds per call))
(using CUDNN_CONVOLUTION_FWD_ALGO_FFT[T] (timed)(cache), ws:37502976, hash:FWD|GPU#0000:01:00.0 -t -g 1 -dim 12,4,256,256,262144,65536,256,1 -filt 5,4,32,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1)
        (ran fwd algo in 0 milliseconds)
        (ran 10 fwd computations in 16 milliseconds (average: 1.6 milliseconds per call))
(using CUDNN_CONVOLUTION_FWD_ALGO_FFT[T] (timed)(cache), ws:37502976, hash:FWD|GPU#0000:01:00.0 -t -g 1 -dim 12,4,256,256,262144,65536,256,1 -filt 5,4,32,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1)
        (ran fwd algo in 0 milliseconds)
        (ran 11 fwd computations in 16 milliseconds (average: 1.45455 milliseconds per call))
ok
theano.gpuarray.tests.test_dnn.TestDnnConv2DRuntimeAlgorithmsWithBigInputs.test_fwd_runtime_algorithms('guess_on_shape_change',) ... (using CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM , ws:65536, hash:FWD|GPU#0000:01:00.0 -g 1 -dim 12,4,128,128,65536,16384,128,1 -filt 5,4,64,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1 [unaligned])
        (selected best suited fwd algo in 0 milliseconds)
        (ran fwd algo in 78 milliseconds)
(using CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM (cache), ws:65536, hash:FWD|GPU#0000:01:00.0 -g 1 -dim 12,4,128,128,65536,16384,128,1 -filt 5,4,64,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1 [unaligned])
        (ran fwd algo in 79 milliseconds)
        (ran 2 fwd computations in 157 milliseconds (average: 78.5 milliseconds per call))
(using CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM (cache), ws:65536, hash:FWD|GPU#0000:01:00.0 -g 1 -dim 12,4,128,128,65536,16384,128,1 -filt 5,4,64,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1 [unaligned])
        (ran fwd algo in 78 milliseconds)
        (ran 3 fwd computations in 235 milliseconds (average: 78.3333 milliseconds per call))
(using CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM (cache), ws:65536, hash:FWD|GPU#0000:01:00.0 -g 1 -dim 12,4,128,128,65536,16384,128,1 -filt 5,4,64,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1 [unaligned])
        (ran fwd algo in 62 milliseconds)
        (ran 4 fwd computations in 297 milliseconds (average: 74.25 milliseconds per call))
(using CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM (cache), ws:65536, hash:FWD|GPU#0000:01:00.0 -g 1 -dim 12,4,128,128,65536,16384,128,1 -filt 5,4,64,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1 [unaligned])
        (ran fwd algo in 79 milliseconds)
        (ran 5 fwd computations in 376 milliseconds (average: 75.2 milliseconds per call))
(using CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM , ws:32768, hash:FWD|GPU#0000:01:00.0 -g 1 -dim 12,4,256,256,262144,65536,256,1 -filt 5,4,32,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1)
        (selected best suited fwd algo in 0 milliseconds)
        (selected 2 fwd algos in 0 milliseconds (average: 0 milliseconds per selection))
        (ran fwd algo in 328 milliseconds)
        (ran 6 fwd computations in 704 milliseconds (average: 117.333 milliseconds per call))
(using CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM , ws:32768, hash:FWD|GPU#0000:01:00.0 -g 1 -dim 12,4,256,256,262144,65536,256,1 -filt 5,4,32,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1 [unaligned])
        (selected best suited fwd algo in 0 milliseconds)
        (selected 3 fwd algos in 0 milliseconds (average: 0 milliseconds per selection))
        (ran fwd algo in 328 milliseconds)
        (ran 7 fwd computations in 1032 milliseconds (average: 147.429 milliseconds per call))
(using CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM (cache), ws:32768, hash:FWD|GPU#0000:01:00.0 -g 1 -dim 12,4,256,256,262144,65536,256,1 -filt 5,4,32,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1 [unaligned])
        (ran fwd algo in 343 milliseconds)
        (ran 8 fwd computations in 1375 milliseconds (average: 171.875 milliseconds per call))
(using CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM (cache), ws:32768, hash:FWD|GPU#0000:01:00.0 -g 1 -dim 12,4,256,256,262144,65536,256,1 -filt 5,4,32,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1 [unaligned])
        (ran fwd algo in 344 milliseconds)
        (ran 9 fwd computations in 1719 milliseconds (average: 191 milliseconds per call))
(using CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM (cache), ws:32768, hash:FWD|GPU#0000:01:00.0 -g 1 -dim 12,4,256,256,262144,65536,256,1 -filt 5,4,32,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1 [unaligned])
        (ran fwd algo in 343 milliseconds)
        (ran 10 fwd computations in 2062 milliseconds (average: 206.2 milliseconds per call))
(using CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM (cache), ws:32768, hash:FWD|GPU#0000:01:00.0 -g 1 -dim 12,4,256,256,262144,65536,256,1 -filt 5,4,32,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1 [unaligned])
        (ran fwd algo in 344 milliseconds)
        (ran 11 fwd computations in 2406 milliseconds (average: 218.727 milliseconds per call))
ok

----------------------------------------------------------------------
Ran 4 tests in 83.132s

OK
```
```
(python2) C:\Users\notoraptor\mila\dev\git\theano>echo %THEANO_FLAGS%
cmodule.debug=True
```
```
(python2) C:\Users\notoraptor\mila\dev\git\theano>echo %CUDA_LAUNCH_BLOCKING%
1
```

Could you do a first review to check if I am on a right way ? @nouiz @abergeron

nouiz commented on theano/gpuarray/c_code/dnn_fwd.c:
```diff
@@ -358,53 +395,72 @@ APPLY_SPECIFIC(conv_fwd)(PyGpuArrayObject *input, PyGpuArrayObject *kerns,
+  /*
+   * This is less than ideal since we need to free it after (which
+   * introduces a synchronization point. But we don't have a module
+   * to place a nice get_work_mem() function in.
```
@notoraptor do you understand this comment?

We cache by default the call to cudaMalloc, so it don't create a real sync. It could create a sync if we where computing many ops in parallel, but currently we put everything on one stream. So this don't currently create any sync point.

@borisfom is it you that wrote that comment?

notoraptor replied:
@nouiz This comment seems to have been introduced in 2015 ( 4d8e60e77df8185da435f7dd793434173709dc76 ) !

nouiz replied:
ok, so can you just remove it? @abergeron can you confirm it is outdated?

abergeron replied:
Yes it is.

nouiz commented on theano/gpuarray/c_code/dnn_fwd.c:
```diff
@@ -358,53 +395,72 @@ APPLY_SPECIFIC(conv_fwd)(PyGpuArrayObject *input, PyGpuArrayObject *kerns,
-    cuda_record(kerns->ga.data, GPUARRAY_CUDA_WAIT_READ);
-    cuda_record((*output)->ga.data, GPUARRAY_CUDA_WAIT_WRITE);
-  }
+  cuda_record(input->ga.data, GPUARRAY_CUDA_WAIT_READ);
```
should we do a record on the workspace too? @abergeron what do you think about that?

Currently, this don't cause problems. But when we work on the multi-stream, it could be needed.

abergeron replied:
We need to do the record before the release otherwise we could have problem in multi-stream.

Also, we need to do a record on all buffers that we touch for proper bookkeeping.

nouiz commented on theano/gpuarray/c_code/dnn_fwd.c:
```diff
@@ -358,53 +395,72 @@ APPLY_SPECIFIC(conv_fwd)(PyGpuArrayObject *input, PyGpuArrayObject *kerns,
+  cuda_wait((*output)->ga.data, GPUARRAY_CUDA_WAIT_WRITE);
+
+  #ifdef DEBUG
+  t = theano_clock();
```
If the user do not enable CUDA_LAUNCH_BLOCKING=1, it will have wrong results. I see two possible fix:
1) Do as with profile=True and test if CUDA_LAUNCH_BLOCKING=1 when cmodule.debug is True
2) make a sync inside the ifdef.

I prefer the second fix as it won't force the sync of all the computation. This would request a sync before taking the first call to theano_clock and another one before the second call for each pair of call.

borisfom replied:
Right - wait should definitely be under #ifdef.

borisfom commented on theano/gpuarray/c_code/dnn_conv_base.c:
```diff
@@ -85,6 +85,23 @@ typedef std::unordered_map<std::string, AlgoRec> AlgoCache;
+#define theano_clock_t std::chrono::time_point
+#define theano_clock() std::chrono::steady_clock::now()
+#define theano_clock_to_milliseconds(t) ( std::chrono::duration_cast<std::chrono::nanoseconds>(t).count() / 1000000.0 )
+#define theano_clock_average_to_milliseconds(t, n) ( theano_clock_to_milliseconds(t) / (n) )
```
This is probably where you lose fractions of milliseconds. 
I was using this successfully to get fractional milliseconds: 
duration.count()*1000.0

borisfom replied:
I was using system_clock though:
 system_clock::time_point last_time = system_clock::now();
....
duration<double> conversion_time(system_clock::now() - current_time);
            printf ("Conversion time is %g ms\n", conversion_time.count()*1000.0);

notoraptor replied:
Indeed, it is more accurate with chrono standard library.

In fact, my compiler was not compiling in C++11 by default, so it used the C interface based on `<ctime>`, which appears to be clearly less accurate.

Therefore, it might be a little hard to get good profilings with C++ < 2011 version.


I'll update my code tomorrow with some fixes needed to make everything work at least in C++11.

nouiz replied:
I'm fine to print the timing only in C++11. This is not something that need to work for everybody everytimes.

borisfom replied:
Codewise is should be straightforward - my initial cut on the cache was in C++11, them I ported it down (unordered map). There are some portability issue with gcc versions - newer gcc would break on c++11 flag so the configure part needs to be there.

notoraptor commented:
Updated.

I tried to support chrono for C++ prior to C++11 by using `<tr1/chrono>` as @borisfom uses `<tr1/unordered_map>`, but `<tr1/chrono>` is not found by g++ (version 4.9.4 on Ubuntu). So, for C++ before 2011, I used instead a third-party library called `plf::nanotimer` wich provides a cross-platform high-precision timing class: http://www.plflib.org/nanotimer.htm (github repo: https://github.com/mattreecebentley/plf_nanotimer ). It's contained into one small header file and licensed under ZLIB license.

What do you think ? If it's right, I can apply changes to the three computations (currently only in dnn_fwd).

Here is an example of output:
```
THEANO_FLAGS=cmodule.debug=True nosetests -xvs theano/gpuarray/tests/test_dnn.py:TestDnnConv2DRuntimeAlgorithmsWithBigInputs.test_fwd_runtime_algorithms > before-c11.log 2>&1
```
NB: This test runs one shape 5 times for `time_once|guess_once` algorithms, and 2 shapes multiple times (total 11 runs) for `time_on_shape_change|guess_on_shape_change` algorithms:

```
Using cuDNN version 7002 on context None
Mapped name None to device cuda: GeForce GTX 1050 (0000:01:00.0)
theano.gpuarray.tests.test_dnn.TestDnnConv2DRuntimeAlgorithmsWithBigInputs.test_fwd_runtime_algorithms('time_once',) ... Using plf_nanotimer: http://www.plflib.org/nanotimer.htm
(using CUDNN_CONVOLUTION_FWD_ALGO_FFT (timed), ws:9445376, hash:FWD|GPU#0000:01:00.0 -t -g 1 -dim 12,4,128,128,65536,16384,128,1 -filt 5,4,64,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1 -ff [unaligned])
	(selected fastest fwd algo in 256.733 milliseconds)
	(ran fwd algo in 0.190667 milliseconds)
Using plf_nanotimer: http://www.plflib.org/nanotimer.htm
(using CUDNN_CONVOLUTION_FWD_ALGO_FFT (timed)(reused), ws:9445376, hash:)
	(ran fwd algo in 0.064351 milliseconds)
	(ran 2 fwd computations in 0.255018 milliseconds (average: 0.127509 milliseconds per call))
Using plf_nanotimer: http://www.plflib.org/nanotimer.htm
(using CUDNN_CONVOLUTION_FWD_ALGO_FFT (timed)(reused), ws:9445376, hash:)
	(ran fwd algo in 0.061178 milliseconds)
	(ran 3 fwd computations in 0.316196 milliseconds (average: 0.105399 milliseconds per call))
Using plf_nanotimer: http://www.plflib.org/nanotimer.htm
(using CUDNN_CONVOLUTION_FWD_ALGO_FFT (timed)(reused), ws:9445376, hash:)
	(ran fwd algo in 0.062092 milliseconds)
	(ran 4 fwd computations in 0.378288 milliseconds (average: 0.094572 milliseconds per call))
Using plf_nanotimer: http://www.plflib.org/nanotimer.htm
(using CUDNN_CONVOLUTION_FWD_ALGO_FFT (timed)(reused), ws:9445376, hash:)
	(ran fwd algo in 0.067904 milliseconds)
	(ran 5 fwd computations in 0.446192 milliseconds (average: 0.0892384 milliseconds per call))
ok
theano.gpuarray.tests.test_dnn.TestDnnConv2DRuntimeAlgorithmsWithBigInputs.test_fwd_runtime_algorithms('guess_once',) ... Using plf_nanotimer: http://www.plflib.org/nanotimer.htm
(using CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM , ws:65536, hash:FWD|GPU#0000:01:00.0 -g 1 -dim 12,4,128,128,65536,16384,128,1 -filt 5,4,64,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1 -ff)
	(selected best suited fwd algo in 0.016507 milliseconds)
	(ran fwd algo in 0.040239 milliseconds)
Using plf_nanotimer: http://www.plflib.org/nanotimer.htm
(using CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM (reused), ws:65536, hash:)
	(ran fwd algo in 0.042181 milliseconds)
	(ran 2 fwd computations in 0.08242 milliseconds (average: 0.04121 milliseconds per call))
Using plf_nanotimer: http://www.plflib.org/nanotimer.htm
(using CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM (reused), ws:65536, hash:)
	(ran fwd algo in 0.032347 milliseconds)
	(ran 3 fwd computations in 0.114767 milliseconds (average: 0.0382557 milliseconds per call))
Using plf_nanotimer: http://www.plflib.org/nanotimer.htm
(using CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM (reused), ws:65536, hash:)
	(ran fwd algo in 0.031538 milliseconds)
	(ran 4 fwd computations in 0.146305 milliseconds (average: 0.0365762 milliseconds per call))
Using plf_nanotimer: http://www.plflib.org/nanotimer.htm
(using CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM (reused), ws:65536, hash:)
	(ran fwd algo in 0.03162 milliseconds)
	(ran 5 fwd computations in 0.177925 milliseconds (average: 0.035585 milliseconds per call))
ok
theano.gpuarray.tests.test_dnn.TestDnnConv2DRuntimeAlgorithmsWithBigInputs.test_fwd_runtime_algorithms('time_on_shape_change',) ... Using plf_nanotimer: http://www.plflib.org/nanotimer.htm
(using CUDNN_CONVOLUTION_FWD_ALGO_FFT (timed)(cache), ws:9445376, hash:FWD|GPU#0000:01:00.0 -t -g 1 -dim 12,4,128,128,65536,16384,128,1 -filt 5,4,64,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1 -ff [unaligned])
	(ran fwd algo in 0.061488 milliseconds)
Using plf_nanotimer: http://www.plflib.org/nanotimer.htm
(using CUDNN_CONVOLUTION_FWD_ALGO_FFT (timed)(cache), ws:9445376, hash:FWD|GPU#0000:01:00.0 -t -g 1 -dim 12,4,128,128,65536,16384,128,1 -filt 5,4,64,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1 -ff [unaligned])
	(ran fwd algo in 0.062292 milliseconds)
	(ran 2 fwd computations in 0.12378 milliseconds (average: 0.06189 milliseconds per call))
Using plf_nanotimer: http://www.plflib.org/nanotimer.htm
(using CUDNN_CONVOLUTION_FWD_ALGO_FFT (timed)(cache), ws:9445376, hash:FWD|GPU#0000:01:00.0 -t -g 1 -dim 12,4,128,128,65536,16384,128,1 -filt 5,4,64,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1 -ff [unaligned])
	(ran fwd algo in 0.061992 milliseconds)
	(ran 3 fwd computations in 0.185772 milliseconds (average: 0.061924 milliseconds per call))
Using plf_nanotimer: http://www.plflib.org/nanotimer.htm
(using CUDNN_CONVOLUTION_FWD_ALGO_FFT (timed)(cache), ws:9445376, hash:FWD|GPU#0000:01:00.0 -t -g 1 -dim 12,4,128,128,65536,16384,128,1 -filt 5,4,64,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1 -ff [unaligned])
	(ran fwd algo in 0.067041 milliseconds)
	(ran 4 fwd computations in 0.252813 milliseconds (average: 0.0632033 milliseconds per call))
Using plf_nanotimer: http://www.plflib.org/nanotimer.htm
(using CUDNN_CONVOLUTION_FWD_ALGO_FFT (timed)(cache), ws:9445376, hash:FWD|GPU#0000:01:00.0 -t -g 1 -dim 12,4,128,128,65536,16384,128,1 -filt 5,4,64,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1 -ff [unaligned])
	(ran fwd algo in 0.060483 milliseconds)
	(ran 5 fwd computations in 0.313296 milliseconds (average: 0.0626592 milliseconds per call))
Using plf_nanotimer: http://www.plflib.org/nanotimer.htm
(using CUDNN_CONVOLUTION_FWD_ALGO_FFT (timed), ws:37502976, hash:FWD|GPU#0000:01:00.0 -t -g 1 -dim 12,4,256,256,262144,65536,256,1 -filt 5,4,32,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1 -ff)
	(selected fastest fwd algo in 1020.92 milliseconds)
	(ran fwd algo in 0.225074 milliseconds)
	(ran 6 fwd computations in 0.53837 milliseconds (average: 0.0897283 milliseconds per call))
Using plf_nanotimer: http://www.plflib.org/nanotimer.htm
(using CUDNN_CONVOLUTION_FWD_ALGO_FFT (timed), ws:37502976, hash:FWD|GPU#0000:01:00.0 -t -g 1 -dim 12,4,256,256,262144,65536,256,1 -filt 5,4,32,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1 -ff [unaligned])
	(selected fastest fwd algo in 1013.75 milliseconds)
	(selected 2 fwd algos in 2034.68 milliseconds (average: 1017.34 milliseconds per selection))
	(ran fwd algo in 0.22298 milliseconds)
	(ran 7 fwd computations in 0.76135 milliseconds (average: 0.108764 milliseconds per call))
Using plf_nanotimer: http://www.plflib.org/nanotimer.htm
(using CUDNN_CONVOLUTION_FWD_ALGO_FFT (timed)(cache), ws:37502976, hash:FWD|GPU#0000:01:00.0 -t -g 1 -dim 12,4,256,256,262144,65536,256,1 -filt 5,4,32,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1 -ff [unaligned])
	(ran fwd algo in 0.064387 milliseconds)
	(ran 8 fwd computations in 0.825737 milliseconds (average: 0.103217 milliseconds per call))
Using plf_nanotimer: http://www.plflib.org/nanotimer.htm
(using CUDNN_CONVOLUTION_FWD_ALGO_FFT (timed)(cache), ws:37502976, hash:FWD|GPU#0000:01:00.0 -t -g 1 -dim 12,4,256,256,262144,65536,256,1 -filt 5,4,32,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1 -ff [unaligned])
	(ran fwd algo in 0.062067 milliseconds)
	(ran 9 fwd computations in 0.887804 milliseconds (average: 0.0986449 milliseconds per call))
Using plf_nanotimer: http://www.plflib.org/nanotimer.htm
(using CUDNN_CONVOLUTION_FWD_ALGO_FFT (timed)(cache), ws:37502976, hash:FWD|GPU#0000:01:00.0 -t -g 1 -dim 12,4,256,256,262144,65536,256,1 -filt 5,4,32,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1 -ff [unaligned])
	(ran fwd algo in 0.066069 milliseconds)
	(ran 10 fwd computations in 0.953873 milliseconds (average: 0.0953873 milliseconds per call))
Using plf_nanotimer: http://www.plflib.org/nanotimer.htm
(using CUDNN_CONVOLUTION_FWD_ALGO_FFT (timed)(cache), ws:37502976, hash:FWD|GPU#0000:01:00.0 -t -g 1 -dim 12,4,256,256,262144,65536,256,1 -filt 5,4,32,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1 -ff [unaligned])
	(ran fwd algo in 0.06133 milliseconds)
	(ran 11 fwd computations in 1.0152 milliseconds (average: 0.0922912 milliseconds per call))
ok
theano.gpuarray.tests.test_dnn.TestDnnConv2DRuntimeAlgorithmsWithBigInputs.test_fwd_runtime_algorithms('guess_on_shape_change',) ... Using plf_nanotimer: http://www.plflib.org/nanotimer.htm
(using CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM (cache), ws:65536, hash:FWD|GPU#0000:01:00.0 -g 1 -dim 12,4,128,128,65536,16384,128,1 -filt 5,4,64,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1 -ff)
	(ran fwd algo in 0.036875 milliseconds)
Using plf_nanotimer: http://www.plflib.org/nanotimer.htm
(using CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM (cache), ws:65536, hash:FWD|GPU#0000:01:00.0 -g 1 -dim 12,4,128,128,65536,16384,128,1 -filt 5,4,64,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1 -ff)
	(ran fwd algo in 0.037617 milliseconds)
	(ran 2 fwd computations in 0.074492 milliseconds (average: 0.037246 milliseconds per call))
Using plf_nanotimer: http://www.plflib.org/nanotimer.htm
(using CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM (cache), ws:65536, hash:FWD|GPU#0000:01:00.0 -g 1 -dim 12,4,128,128,65536,16384,128,1 -filt 5,4,64,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1 -ff)
	(ran fwd algo in 0.031896 milliseconds)
	(ran 3 fwd computations in 0.106388 milliseconds (average: 0.0354627 milliseconds per call))
Using plf_nanotimer: http://www.plflib.org/nanotimer.htm
(using CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM (cache), ws:65536, hash:FWD|GPU#0000:01:00.0 -g 1 -dim 12,4,128,128,65536,16384,128,1 -filt 5,4,64,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1 -ff)
	(ran fwd algo in 0.031589 milliseconds)
	(ran 4 fwd computations in 0.137977 milliseconds (average: 0.0344943 milliseconds per call))
Using plf_nanotimer: http://www.plflib.org/nanotimer.htm
(using CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM (cache), ws:65536, hash:FWD|GPU#0000:01:00.0 -g 1 -dim 12,4,128,128,65536,16384,128,1 -filt 5,4,64,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1 -ff)
	(ran fwd algo in 0.031219 milliseconds)
	(ran 5 fwd computations in 0.169196 milliseconds (average: 0.0338392 milliseconds per call))
Using plf_nanotimer: http://www.plflib.org/nanotimer.htm
(using CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM , ws:32768, hash:FWD|GPU#0000:01:00.0 -g 1 -dim 12,4,256,256,262144,65536,256,1 -filt 5,4,32,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1 -ff)
	(selected best suited fwd algo in 0.007877 milliseconds)
	(ran fwd algo in 0.034566 milliseconds)
	(ran 6 fwd computations in 0.203762 milliseconds (average: 0.0339603 milliseconds per call))
Using plf_nanotimer: http://www.plflib.org/nanotimer.htm
(using CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM (cache), ws:32768, hash:FWD|GPU#0000:01:00.0 -g 1 -dim 12,4,256,256,262144,65536,256,1 -filt 5,4,32,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1 -ff)
	(ran fwd algo in 0.038053 milliseconds)
	(ran 7 fwd computations in 0.241815 milliseconds (average: 0.034545 milliseconds per call))
Using plf_nanotimer: http://www.plflib.org/nanotimer.htm
(using CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM (cache), ws:32768, hash:FWD|GPU#0000:01:00.0 -g 1 -dim 12,4,256,256,262144,65536,256,1 -filt 5,4,32,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1 -ff)
	(ran fwd algo in 0.031602 milliseconds)
	(ran 8 fwd computations in 0.273417 milliseconds (average: 0.0341771 milliseconds per call))
Using plf_nanotimer: http://www.plflib.org/nanotimer.htm
(using CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM (cache), ws:32768, hash:FWD|GPU#0000:01:00.0 -g 1 -dim 12,4,256,256,262144,65536,256,1 -filt 5,4,32,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1 -ff)
	(ran fwd algo in 0.032044 milliseconds)
	(ran 9 fwd computations in 0.305461 milliseconds (average: 0.0339401 milliseconds per call))
Using plf_nanotimer: http://www.plflib.org/nanotimer.htm
(using CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM (cache), ws:32768, hash:FWD|GPU#0000:01:00.0 -g 1 -dim 12,4,256,256,262144,65536,256,1 -filt 5,4,32,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1 -ff)
	(ran fwd algo in 0.031919 milliseconds)
	(ran 10 fwd computations in 0.33738 milliseconds (average: 0.033738 milliseconds per call))
Using plf_nanotimer: http://www.plflib.org/nanotimer.htm
(using CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM (cache), ws:32768, hash:FWD|GPU#0000:01:00.0 -g 1 -dim 12,4,256,256,262144,65536,256,1 -filt 5,4,32,64 -mode conv -pad 0,0 -subsample 1,1 -dilation 1,1 -ff)
	(ran fwd algo in 0.032618 milliseconds)
	(ran 11 fwd computations in 0.369998 milliseconds (average: 0.0336362 milliseconds per call))
ok

----------------------------------------------------------------------
Ran 4 tests in 61.958s

OK
```

@nouiz @abergeron @lamblin

nouiz commented:
I don't like adding code with another license. It complicate the licensing. Why not just don't support that case? We could just print `no timing, c++11 needed` or something like that.

What do you think of that option?

notoraptor commented:
Updated. Removed third-party license, so profiling is available only with C++11 and later.

Profiling added to gradweight and gradinput too, so this PR is ready to review.

@nouiz @abergeron

notoraptor commented:
jenkins test this

borisfom commented:
Wait, I was actually down-porting this code originally so that it does not require C++11 (which I am all for anyway). If you do a simple conditional, you'll be fine (look up exact constant for c++11 standard):
#if __cplusplus < 2011CONSTANT
#  include <chrono>
# include <unordered_map>
# else
#  include <chrono>
# include <unordered_map>
#endif

notoraptor commented:
@borisfom Yes, I have tried the same condition for `<chrono>` with `<tr1/chrono>`, but my compiler can not find `<tr1/chrono>` when C++11 is not activated.

notoraptor commented:
With this file:
**`test.cpp`**
```C++
#include <tr1/unordered_map>
#include <tr1/chrono>

int main() {
	return 0;
}
```
When I compile:
```
g++ test.cpp -o test
```
```
$ g++ test.cpp -o test
test.cpp:2:22: fatal error: tr1/chrono: Aucun fichier ou dossier de ce type
 #include <tr1/chrono>
                      ^
compilation terminated.
```
It finds `<tr1/unordered_map>` but not `<tr1/chrono>`. It seems `<chrono>` is not in `tr1`.
My g++ version:
```
$ g++ -v
Using built-in specs.
COLLECT_GCC=g++
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/4.9/lto-wrapper
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 4.9.4-2ubuntu1' --with-bugurl=file:///usr/share/doc/gcc-4.9/README.Bugs --enable-languages=c,c++,java,go,d,fortran,objc,obj-c++ --prefix=/usr --program-suffix=-4.9 --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --with-gxx-include-dir=/usr/include/c++/4.9 --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --enable-gnu-unique-object --disable-vtable-verify --enable-plugin --with-system-zlib --disable-browser-plugin --enable-java-awt=gtk --enable-gtk-cairo --with-java-home=/usr/lib/jvm/java-1.5.0-gcj-4.9-amd64/jre --enable-java-home --with-jvm-root-dir=/usr/lib/jvm/java-1.5.0-gcj-4.9-amd64 --with-jvm-jar-dir=/usr/lib/jvm-exports/java-1.5.0-gcj-4.9-amd64 --with-arch-directory=amd64 --with-ecj-jar=/usr/share/java/eclipse-ecj.jar --enable-objc-gc --enable-multiarch --disable-werror --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 4.9.4 (Ubuntu 4.9.4-2ubuntu1)
```

borisfom commented:
@notoraptor : right, <chrono> was never part of TR1, my bad. Guarding the timing code out is the right thing then,

nouiz commented on theano/gpuarray/c_code/dnn_fwd.c:
```diff
@@ -282,10 +314,16 @@ APPLY_SPECIFIC(conv_fwd)(PyGpuArrayObject *input, PyGpuArrayObject *kerns,
 #endif
       } else {
+        #ifdef DEBUG_TIMING
+        timer.start();
```
Can you confirm that this reset the timer?

notoraptor replied:
Yes, it resets the timer.

nouiz commented on theano/gpuarray/c_code/dnn_gi.c:
```diff
@@ -159,11 +185,12 @@ APPLY_SPECIFIC(conv_gi)(PyGpuArrayObject *kerns, PyGpuArrayObject *output,
   cuda_enter(c->ctx);
 
+  size_t maxfree = c_get_largest_free_block_size(c);
+  if (PyErr_Occurred()) return 1;
```
We would need to exit the context. I think this is the reason why it was before cuda_enter()

notoraptor replied:
Indeed. I changed the place to have same code layout as in `dnn_fwd`, but even in `dnn_fwd` we forget to exit the context.

I will exit context properly in all 3 files.

nouiz commented on theano/gpuarray/c_code/dnn_gw.c:
```diff
@@ -146,11 +172,12 @@ APPLY_SPECIFIC(conv_gw)(PyGpuArrayObject *input, PyGpuArrayObject *output,
 
   cuda_enter(c->ctx);
 
+  size_t maxfree = c_get_largest_free_block_size(c);
```
idem

notoraptor replied:
Fixed

nouiz commented on theano/gpuarray/c_code/dnn_gw.c:
```diff
@@ -305,6 +348,17 @@ APPLY_SPECIFIC(conv_gw)(PyGpuArrayObject *input, PyGpuArrayObject *output,
      );
 #endif
+#ifdef DEBUG_TIMING
+    if (!(reuse_algo || use_cached)) {
```
Here, if we always reuse one shape, we will never print the timing of computation after the first call. What about printing it for every n_section%100? or  1000 or something like that?

Ideally, it would be to print  it at the last call, but I don't know how to do that.

Do you think this is overkill and it won't be used?

notoraptor replied:
This code block prints timing info about algorithm selection, thus if we always reuse one shape, it will be printed at the first call (when selection is done) but not on other calls (as there is no selection in next calls), to avoid overload the console output.

Timing info about computation itself will be always printed (see lines below just before the final `return 0;`).

Is it acceptable ?

nouiz replied:
I would go with that for now. Maybe we could n_selection and/or the selection time to the computation line. We will see.

notoraptor commented:
@nouiz updated.
